libraries/ed25519-donna/bech32/bech32.1
.\" By nullius <nullius@nym.zone>
.\" PGP:	0xC2E91CD74A4C57A105F6C21B5A00591B2F307E0C
.\" Bitcoin:	3NULL3ZCUXr7RDLxXeLPDMZDZYxuaYkCnG
.\"		bc1qcash96s5jqppzsp8hy8swkggf7f6agex98an7h
.\"
.\" Copyright (c) 2017.  All rights reserved.
.\"
.\" The Antiviral License (AVL) v0.0.1, with added Bitcoin Consensus Clause:
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of the source code must retain the above copyright
.\"    and credit notices, this list of conditions, and the following
.\"    disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    and credit notices, this list of conditions, and the following
.\"    disclaimer in the documentation and/or other materials provided
.\"    with the distribution.
.\" 3. Derivative works hereof MUST NOT be redistributed under any license
.\"    containing terms which require derivative works and/or usages to
.\"    publish source code, viz. what is commonly known as a "copyleft"
.\"    or "viral" license.
.\" 4. Derivative works hereof which have any functionality related to
.\"    digital money (so-called "cryptocurrency") MUST EITHER adhere to
.\"    consensus rules fully compatible with Bitcoin Core, OR use a name
.\"    which does not contain the word "Bitcoin".
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd December 31, 2017
.Dt BECH32 1
.Os Bitcoin
.Sh NAME
.Nm bech32
.Nd "Encode and decode Bech32 strings and Bitcoin Segwit addresses"
.Sh SYNOPSIS
General-purpose encoding and decoding:
.Pp
.Nm
.Fl e
.Fl h Ar hrp
.Ar hexdata
.Nm
.Fl e
.Fl 5
.Fl h Ar hrp
.Ar base32data
.Nm
.Fl d
.Op Fl 5
.Op Fl u
.Ar bech32string
.Pp
For Bitcoin
.Sy "Bravo Charlie"
addresses
.Pq Dq bc1 :
.Pp
.Nm
.Fl s Ar witver
.Ar hexdata
.Nm
.Fl S
.Ar bc1addr
.Pp
For .onion domains:
.Pp
.Nm
.Fl e
.Ar base32domain.onion
.Nm
.Fl d
.Ar onion1bech32
.Sh DESCRIPTION
The
.Nm
utility is a powerful general-purpose tool for encoding and decoding
of BIP 173 standard Bech32 strings and Bitcoin Bech32 addresses.
.Pp
Hexadecimal data may be input with or without a leading
.Dq 0x
prefix.
It is output with such a prefix, to prevent ambiguity.
.Pp
It has the following modes, some of which overlap:
.Bl -tag -width ".Fl d Ar argument"
.It Fl e
Encoder mode.
Output is a Bech32 string.
.It Fl d
Decoder mode.
Output is the input string
.Ap
s human-readable part
.Pq HRP ,
followed by a delimiting colon
.Pq Dq ":" ,
followed by the
.Dq "data part"
data in hexadecimal with a
.Dq 0x
prefix.
.It Fl s Ar witver
Encode a Bitcoin Segwit address with witness version
.Ar witver .
.It Fl S
Decode a Bitcoin Segwit address.
Output its witness version in
.Em decimal
.Pq "0\(en16 inclusive" ,
followed by a delimiting colon
.Pq Dq ":" ,
followed by the address data in
.Em hexadecimal
with a
.Dq 0x
prefix.
.It Pq ".onion autodetect"
Dot-onion mode.
When encoding,
.Nm
will automatically detect a dot-onion domain and read its second level
as RFC 4648 Base32 data.
When decoding,
.Nm
will detect a
.Dq onion
HRP, and output a dot-onion domain.
.El
.Pp
The options are as follows:
.Bl -tag -width ".Fl d Ar argument"
.It Fl h Ar hrp
General encoding only, and required therefor.
Provide the Human-Readable Portion
.Pq HRP
for the Bech32 string.
.It Fl 5
Read RFC 4648 Base32 data when encoding, or write RFC 4648 Base32 data
when decoding.
.It Fl u
.Pq "Decoding only."
.\" XXX: What about Base32?
Output hexadecimal characters in uppercase.
.El
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
Extract the witness version and Hash160 from the
.Nm
utility author
.Ap
s Bech32 tip address:
.Pp
.Dl bech32 -S bc1qcash96s5jqppzsp8hy8swkggf7f6agex98an7h
.Dl 0:0xc76172ea149002114027b90f0759084f93aea326
.Pp
Get a
.Dq "hello, world"
introduction to Bech32:
.Pp
.Dl bech32 -e -h hello_world 48656c6c6f2c20776f726c6421
.Dl hello_world1fpjkcmr09ss8wmmjd3jzzwhs4ff
.Pp
Generate a
.Dq "burn address"
with a Hash160 of all zeroes, which would be spendable by the same unknown
private keys as the infamous 1111111111111111111114oLvT2.
.Sy "Warning:  Do NOT send coins here:"
.Pp
.Dl bech32 -s 0 0x0000000000000000000000000000000000000000
.Dl bc1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq9e75rs
.Pp
Bech32-encode the address for Wikileaks, to add error-correcting codes:
.Pp
.Dl bech32 -e wlupld3ptjvsgwqw.onion
.Dl onion1kt50trm0nf4jxkskpcjy74
.Pp
Now, decode the address someone gave you:
.Pp
.Dl bech32 -d onion1kt50trm0nf4jxkskpcjy74
.Dl wlupld3ptjvsgwqw.onion
.Sh STANDARDS
The
.Nm
utility conforms to:
.Bl -bullet
.It
The Bech32 standard:
P. Wuille, G. Maxwell,
.Dq "Base32 address format for native v0\(en16 witness outputs" ,
2016-03-20, Bitcoin
.Lk https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki "BIP 173" .
.It
S. Josefsson,
.Dq "The Base16, Base32, and Base64 Data Encodings" ,
2006-10,
.Lk https://tools.ietf.org/html/rfc4648 "RFC 4648" .
.Pq "For transcoding RFC standard Base32 input data and .onion domains."
.It
J. Appelbaum, A. Muffett,
.Do "The" So .onion Sc "Special-Use Domain Name" Dc ,
2015-10,
.Lk https://tools.ietf.org/html/rfc7686 "RFC 7686" .
.El
.Pp
The term
.Dq Sy "Bravo Charlie Address"
and the motto
.Dq Sy "Bravo Charlie One means money"
are from a
.Lk https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-December/015477.html "proposal by this author" .
.Sh AUTHORS
The
.Nm
utility was written by
.An nullius
.Aq Mt nullius@nym.zone .
.Pp
PGP:
.Li 0xC2E91CD74A4C57A105F6C21B5A00591B2F307E0C
.Pp
Tips:
.Lk bitcoin:3NULL3ZCUXr7RDLxXeLPDMZDZYxuaYkCnG 3NULL3ZCUXr7RDLxXeLPDMZDZYxuaYkCnG ,
.Lk bitcoin:bc1qcash96s5jqppzsp8hy8swkggf7f6agex98an7h bc1qcash96s5jqppzsp8hy8swkggf7f6agex98an7h .
.Pp
The internal Bech32 encoding and decoding is done by the open-source
.Lk https://github.com/sipa/bech32/tree/master/ref/c "Bech32 reference code"
written by Pieter Wuille
.Pq sipa
.Pq "no affiliation with this author" .
.Sh BUGS
This code started as a quickly-made utility for personal use, and kept
growing as such things are wont to do.
The user interface is reasonably logical, given the tool
.Ap
s flexibility.
However, the source code needs some substantial refactoring.
.Pp
This manpage could use more and better examples, plus some other general
editing.
.Pp
Test cases are needed.
Unfortunately, the Bech32 standard does not currently provide full roundtrip
test vectors for arbitrary Bech32 strings.
.Pp
Special support is planned for a concept which this author calls
.Dq Sy "PGP Descriptors" .
However, a spec must be drawn before releasing such a thing into the wild.
.Sh SECURITY CONSIDERATIONS
This is an early release, which should be considered alpha-quality software.
It
.Sy should not
be used on untrusted inputs, such as anything blindly accepted by a webserver.
High on the author
.Ap
s TODO list is to beef up input validation.
At this time, aside from a few simple checks, the utility will happily
pass the buck to the Bech32 reference functions.


libraries/ed25519-donna/bech32/bech32.c
/**
 * By nullius <nullius@nym.zone>
 * PGP:		0xC2E91CD74A4C57A105F6C21B5A00591B2F307E0C
 * Bitcoin:	3NULL3ZCUXr7RDLxXeLPDMZDZYxuaYkCnG
 *		bc1qcash96s5jqppzsp8hy8swkggf7f6agex98an7h
 *
 * Copyright (c) 2017.  All rights reserved.
 *
 * The Antiviral License (AVL) v0.0.1, with added Bitcoin Consensus Clause:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of the source code must retain the above copyright
 *    and credit notices, this list of conditions, and the following
 *    disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    and credit notices, this list of conditions, and the following
 *    disclaimer in the documentation and/or other materials provided
 *    with the distribution.
 * 3. Derivative works hereof MUST NOT be redistributed under any license
 *    containing terms which require derivative works and/or usages to
 *    publish source code, viz. what is commonly known as a "copyleft"
 *    or "viral" license.
 * 4. Derivative works hereof which have any functionality related to
 *    digital money (so-called "cryptocurrency") MUST EITHER adhere to
 *    consensus rules fully compatible with Bitcoin Core, OR use a name
 *    which does not contain the word "Bitcoin".
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef __linux__
#define	_POSIX_C_SOURCE	200809L
#endif

#include <sys/types.h>
#include <unistd.h>

#include <assert.h>
#include <errno.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <err.h>

#include "segwit_addr.h"

#define	MODE_DECODE		0
#define	MODE_ENCODE		1
#define	MODE_SEGWIT_DECODE	2
#define	MODE_SEGWIT_ENCODE	3
#define	MODE_ONION_DECODE	0x00012000
#define	MODE_ONION_ENCODE	0x00013000
#define	IO_HEX			0x10
#define	IO_RFC4648		0x20

static ssize_t
hexdec(unsigned char *data, size_t len, const char *hexdata, size_t hexdata_len)
{
	ssize_t data_len = 0;

	if (len < hexdata_len / 2 || (hexdata_len & 1))
		return (-1);

	for (size_t i = 0, j = 0; i < hexdata_len; i+=2, ++j) {
#ifdef notyet
	/* For copypaste of old-style PGP fingerprint display, etc. */
	Unfortunately, this musses the length checks in main() and
	the loop counting here.  General concept:

	if (isblank(hexdata[i]))
		continue;
#endif
		if (hexdata[i] >= '0' && hexdata[i] <= '9')
			data[j] = (hexdata[i] - '0') << 4;
		else if (hexdata[i] >= 'A' && hexdata[i] <= 'F')
			data[j] = (hexdata[i] - ('A' - 0xa)) << 4;
		else if (hexdata[i] >= 'a' && hexdata[i] <= 'f')
			data[j] = (hexdata[i] - ('a' - 0xa)) << 4;
		else
			return (-1);

		if (hexdata[i+1] >= '0' && hexdata[i+1] <= '9')
			data[j] |= hexdata[i+1] - '0';
		else if (hexdata[i+1] >= 'A' && hexdata[i+1] <= 'F')
			data[j] |= hexdata[i+1] - ('A' - 0xa);
		else if (hexdata[i+1] >= 'a' && hexdata[i+1] <= 'f')
			data[j] |= hexdata[i+1] - ('a' - 0xa);
		else
			return (-1);

		++data_len;
	}

	return (data_len);
}

ssize_t
hexenc(char *hd, size_t buflen, const unsigned char *data, size_t datalen, int cs)
{
	const char hex[2][16] = { "0123456789abcdef", "0123456789ABCDEF" };
	ssize_t hexlen = 0;

	if (buflen < datalen * 2 + 1)
		return (-1);

	cs = !!cs;

	for (int i = 0; i < datalen; ++i) {
		*hd++ = hex[cs][data[i] >> 4];
		*hd++ = hex[cs][data[i] & 0xf];
		hexlen += 2;
	}
	*hd = '\0';

	return (hexlen);
}

static ssize_t
b32enc(unsigned char *b32, size_t b32buflen, const unsigned char *data, size_t datalen)
{
	unsigned bits = 0, b32char = 0;
	ssize_t b32data_len = 0;

	if (b32buflen < (datalen * 8 / 5 + !!(datalen * 8 % 5)))
		return (-1);

	do {
		b32char <<= 8, b32char |= *data++, bits += 8;

		while (bits >= 5) {
			*b32++ = (b32char >> (bits - 5)), ++b32data_len;
			b32char &= ~(0x1f << (bits -= 5));
		}
	} while (--datalen > 0);

	assert(bits < 5);
	if (bits > 0)
		*b32 = b32char << (5 - bits), ++b32data_len;

	return (b32data_len);
}

static ssize_t
b32dec(unsigned char *data, size_t buflen, const unsigned char *b32data, size_t b32datalen)
{
	unsigned bits = 0, u8char = 0;
	ssize_t datalen = 0;

	do {
		if (*b32data > 0x1f)
			return (-1);

		u8char <<= 5, u8char |= *b32data++, bits += 5;
		while (bits >= 8) {
			*data++ = u8char >> (bits - 8), ++datalen;
			u8char &= ~(0xff << (bits -= 8));
		}
	} while (--b32datalen > 0);
	assert(bits <= 8); /* Guaranteed by logic. */
	if (bits > 4 || u8char != 0) /* Invalid per the specification. */
		return (-1);

	return (datalen);
}

static ssize_t
rfcb32enc(char *b32, const unsigned char *data, size_t datalen)
{
	const char alphabet[32] = "abcdefghijklmnopqrstuvwxyz234567";
	ssize_t len = 0;

	while (datalen > 0) {
		if (*data > 31)
			return (-1);
		*b32++ = alphabet[*data++], --datalen, ++len;
	}
	*b32 = '\0';
}

static ssize_t
rfcb32dec(unsigned char *data, size_t buflen, const char *b32, size_t b32len)
{
	ssize_t datalen = 0;

	while (b32len > 0) {
		if (*b32 >= 'a' && *b32 <= 'z')
			*data = *b32 - 'a';
		else if (*b32 >= 'A' && *b32 <= 'Z')
			*data = *b32 - 'A';
		else if (*b32 >= '2' && *b32 <= '7')
			*data = *b32 - ('2' - 26);
		else
			return (-1);

		++b32, ++data, ++datalen, --b32len;
	}

	return (datalen);
}

int
main(int argc, char *argv[])
{
	int ch, error, mode = -1, b32mode = 8, hexcase = 0;
	long witver = -1;
	const char *hrp = NULL, *hexdata = NULL, *bechdata = NULL, *cur;
	char bech32[256], codehrp[84], *str = bech32, *endptr;
	unsigned char data[80], b32data[128], *b32cur = b32data;
	ssize_t hrp_len, data_len = 0, b32data_len = 0, hexdata_len;

	while ((ch = getopt(argc, argv, "58Sdeh:s:u")) > -1) {
		switch (ch) {
		case '5':
			b32mode = 5;
			break;
		case '8': /* default */
			b32mode = 8;
			break;
		case 'S':
			mode = MODE_SEGWIT_DECODE;
			break;
		case 'd':
			mode = MODE_DECODE;
			break;
		case 'e':
			mode = MODE_ENCODE;
			break;
		case 'h':
			hrp = optarg;
			break;
		case 's':
			mode = MODE_SEGWIT_ENCODE;
			errno = 0;
			witver = strtol(optarg, &endptr, 10);
			if (witver < 0 || witver > 16 ||
				*endptr != '\0' || errno != 0)
				witver = -2;
			break;
		case 'u':
			hexcase = 1;
			break;
		default:
			errx(1, "Bad option");
		}
	}
	argc -= optind, argv += optind;

	if (*argv == NULL || *(argv+1) != NULL)
		errx(1, "Wrong options");

	if (witver == -2)
		errx(1, "Invalid witness version provided with -s");

	if (witver != -1 && hrp != NULL)
		errx(1, "-h hrp is automatically set by -s option");

	if (witver != -1 && mode != MODE_SEGWIT_ENCODE)
		errx(1, "Mode conflict");

	switch (mode) {
	case MODE_ENCODE:
		if ((cur = strchr(*argv, '.')) != NULL) {
			if (strcmp(cur, ".onion") != 0)
				errx(1, "Bad encoding string");
			hrp = "onion";
			hrp_len = strlen(hrp);
			hexdata = *argv;
			hexdata_len = cur - *argv;
			b32mode = 5;
			mode = MODE_ONION_ENCODE;
			goto onion_encode;
		} else {
			if (hrp == NULL)
				errx(1, "HRP needed (-h hrp)");
			hrp_len = strlen(hrp);
			if (hrp_len < 1 || hrp_len > 83)
				errx(1, "Bad HRP length");
		}

		/* FALLTHRUOGH */
	case MODE_SEGWIT_ENCODE:
		hexdata = *argv;
		hexdata_len = strlen(hexdata);
		if (hexdata_len < 6 || (hexdata_len & 1))
			errx(1, "Bad hex data length");

		if (strncmp(hexdata, "0x", 2) == 0) {
			hexdata += 2, hexdata_len -= 2;
			if (hexdata_len < 6)
				errx(1, "Bad hex data length");
		}

		/* XXX: This leaves the real check to bech32_encode(). */
		if (hexdata_len > sizeof(data) * 2)
			errx(1, "Bad hex data length");

		if (mode == MODE_SEGWIT_ENCODE)
			goto segwit_encode;

onion_encode:
		if (b32mode == 5) {
			b32data_len = rfcb32dec(b32data, sizeof(b32data), hexdata, hexdata_len);
			if (b32data_len < 0)
				errx(1, "Bad base32 data");
		} else {
			data_len = hexdec(data, sizeof(data), hexdata, hexdata_len);
			if (data_len < 0)
				errx(1, "Bad hex data");

			b32data_len = b32enc(b32data, sizeof(b32data), data, data_len);
			if (b32data_len < 0)
				errx(1, "base32 encoding failed");
		}

		error = bech32_encode(bech32, hrp, b32data, b32data_len);
		if (error != 1)
			errx(1, "bech32_encode() failed (data_len: %ju)",
				(intmax_t)data_len);

		printf("%s\n", bech32);
		break;
segwit_encode:
		data_len = hexdec(data, sizeof(data), hexdata, hexdata_len);
		if (data_len < 0)
			errx(1, "Bad hex data");
fprintf(stderr, "data_len: %jd\n", (intmax_t)data_len);
		error = segwit_addr_encode(bech32, "bc", witver, data, data_len);
		if (error != 1)
			errx(1, "segwit_addr_encode() failed");

		printf("%s\n", bech32);
		break;
	case MODE_DECODE:
		error = bech32_decode(codehrp, b32data, &b32data_len, *argv);
		if (error != 1)
			errx(1, "bech32_decode() failed");

		if (!strcmp(codehrp, "onion"))
			b32mode = 5, mode = MODE_ONION_DECODE;

		if (b32mode == 5) {
			rfcb32enc(str, b32data, b32data_len);
			if (mode == MODE_ONION_DECODE)
				printf("%s.onion\n", str);
			else
				printf("%s:%s\n", codehrp, str);
		} else {
			data_len = b32dec(data, sizeof(data), b32data, b32data_len);
			hexenc(str, sizeof(bech32), data, data_len, hexcase);

			printf("%s:0x%s\n", codehrp, str);
		}
		break;
	case MODE_SEGWIT_DECODE: {
		int ver;

		error = segwit_addr_decode(&ver, data, &data_len, "bc", *argv);
		if (error != 1)
			errx(1, "segwit_addr_decode() failed");

		hexenc(str, sizeof(bech32), data, data_len, hexcase);

		printf("%d:0x%s\n", ver, str);

		break;
	}
	default:
		errx(1, "Mode not set (encode or decode)");
	}

	return (0);
}
